### PR TITLE
Log streaming

### PR DESCRIPTION
**Pull request by aastein**
fixes: #222 

client cli:
```
$ ./akash logs -h
service logs

Usage:
  akash logs <lease> [flags]

Flags:
  -f, --follow        Follow the log stream of the pod
  -h, --help          help for logs
  -l, --lines int     Number of lines from the end of the logs to show per service (default 10)
  -n, --node string   node host (default "http://localhost:46657")

Global Flags:
  -d, --data string   data directory (default "/Users/aaron/.akash")
```

```
ouput:
`[<pod name>] <time> <log contents>`
```

exmaple:
```
$ ./akash logs web 77a3d4d386c5df4169cb1c9943dfdfe394729f1eda2c7711c4b00e72850d9fbf/1/2/8224e14f903a2e136a6362527b19f11935197175cb69981940933aa04459a2a9 -l 1 -f
[web-58c8984dfd-nbf6g]  2018-07-04T01:57:55.141522165Z 172.17.0.5 - - [04/Jul/2018:01:57:55 +0000] "GET /images/favicon.png HTTP/1.1" 200 1825 "http://hello.192.168.99.100.nip.io/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.99 Safari/537.36" "192.168.99.1"
[web-58c8984dfd-nbf6g]  2018-07-04T01:57:57.255819449Z 172.17.0.5 - - [04/Jul/2018:01:57:57 +0000] "GET /images/favicon.png HTTP/1.1" 200 1825 "http://hello.192.168.99.100.nip.io/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.99 Safari/537.36" "192.168.99.1"
[web-58c8984dfd-nbf6g]  2018-07-04T02:03:04.221319604Z 172.17.0.5 - - [04/Jul/2018:02:03:04 +0000] "GET / HTTP/1.1" 304 0 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.99 Safari/537.36" "192.168.99.1"
```

### CODE REVIEW

File: provider/grpc/server_test.go [hunk 1]
```diff
@@ -53,14 +56,75 @@ func TestStatus(t *testing.T) {
...
+func (s mockCluserServiceLogServer) Send(log *types.Log) error {
+	sent = append(sent, log)
+	return nil
+}
+
+func TestServiceLogs(t *testing.T) {
+	handler := new(mmocks.Handler)
+	client := new(kmocks.Client)
+	message := "logs logs logs logs\n"
+	stream := testutil.ReadCloser{bytes.NewBuffer([]byte(message))}
```
Comment (golangci) [inline provider/grpc/server_test.go]: github.com/ovrclk/akash/testutil.ReadCloser composite literal uses unkeyed fields

File: provider/grpc/server.go [hunk 2]
```diff
@@ -70,43 +70,66 @@ func (s server) Deploy(context context.Context, req *types.ManifestRequest) (*ty
...
+
+	for remaining := len(logs); remaining > 0; {
+		select {
+		case err := <-errch:
+			if err != nil {
+				s.log.Error(err.Error())
+			}
+			remaining--
+		case entry := <-logch:
+			server.Send(entry)
```
Comment (golangci) [inline provider/grpc/server.go]: Error return value of `server.Send` is not checked

File: provider/cluster/kube/client.go [hunk 3]
```diff
@@ -162,12 +152,68 @@ func (c *client) Deploy(lid types.LeaseID, group *types.ManifestGroup) error {
...
 
-func (c *client) shouldExpose(expose *types.ManifestServiceExpose) bool {
-	return expose.Global &&
-		(expose.ExternalPort == 80 ||
-			(expose.ExternalPort == 0 && expose.Port == 80))
+func (c *client) TeardownNamespace(ns string) error {
+	return c.kc.CoreV1().Namespaces().Delete(ns, &metav1.DeleteOptions{})
+}
+
+func (c *client) ServiceLogs(ctx context.Context, lid types.LeaseID, tailLines int64, follow bool) ([]*cluster.ServiceLog, error) {
```
Comment (golangci) [inline provider/cluster/kube/client.go]: line is 131 characters

File: provider/cluster/kube/client.go [hunk 4]
```diff
@@ -162,12 +152,68 @@ func (c *client) Deploy(lid types.LeaseID, group *types.ManifestGroup) error {
...
+	if err != nil {
+		c.log.Error(err.Error())
+		return nil, types.ErrInternalError{Message: "internal error"}
+	}
+	if deployments == nil {
+		return nil, types.ErrResourceNotFound{Message: "no deployments for lease"}
+	}
+	response := &types.LeaseStatusResponse{}
+	for _, deployment := range deployments.Items {
+		status := &types.LeaseStatus{Name: deployment.Name, Status: fmt.Sprintf("available replicas: %v/%v", deployment.Status.AvailableReplicas, deployment.Status.Replicas)}
```
Comment (golangci) [inline provider/cluster/kube/client.go]: line is 168 characters

File: provider/grpc/server.go [hunk 5]
```diff
@@ -70,43 +70,66 @@ func (s server) Deploy(context context.Context, req *types.ManifestRequest) (*ty
...
-	deployments, err := s.Client.KubeDeployments(lease.LeaseID)
 	if err != nil {
 		s.log.Error(err.Error())
 		return nil, types.ErrInternalError{Message: "internal error"}
 	}
-	if deployments == nil {
+	return s.Client.LeaseStatus(lease.LeaseID)
+}
+
+func (s server) ServiceStatus(ctx context.Context, req *types.ServiceStatusRequest) (*types.ServiceStatusResponse, error) {
```
Comment (golangci) [inline provider/grpc/server.go]: line is 123 characters

File: cmd/akash/logs.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	options := types.LogOptions{
+		TailLines: tailLines,
+		Follow:    follow,
+	}
+	b, err := json.Marshal(options)
+	if err != nil {
+		return err
+	}
+	url := provider.HostURI + "/logs/" + leasePath + "/" + serviceName
+	body, err := stream(session.Ctx(), url, b)
```
Comment (boz) [inline cmd/akash/logs.go]: To have this shut down gracefully, wrap it in a `common.RunForever()`.  Though as I type this I realize that maybe we should wrap everything in that.

File: types/types.proto [hunk 7]
```diff
@@ -426,6 +432,30 @@ message ServiceStatusResponse {
    int32 availableReplicas  = 5; 
 }
 
+message LogRequest {
+  string name       = 1;
+  string deployment = 2;
+  string group      = 3;
+  string order      = 4;
+  string provider   = 5;
+  LogOptions options = 6;
```
Comment (boz) [inline types/types.proto]: [ahem](http://vimcasts.org/episodes/aligning-text-with-tabular-vim/)

File: cmd/akash/logs.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+	if err != nil {
+		return err
+	}
+
+	scanner := bufio.NewScanner(body)
+	common.RunForever(printLog(session, scanner))
```
Comment (golangci) [inline cmd/akash/logs.go]: Error return value of `common.RunForever` is not checked

**Pull request merged**